Library/OcAppleKernelLib/CommonPatches.c

```diff
@@ -1064,6 +1064,135 @@ PatchIncreasePciBarSize (
   return Status;
 }
 
+STATIC
+CONST UINT8
+mSerialDevicePmioFind[] = {
+  0x66, 0xBA, 0xF8, 0x03  ///< mov dx, 0x03F[8-9A-F]
+};
+
+STATIC
+UINT8
+mSerialDevicePmioPort[] = {
+  0x00, 0x00              ///< To be set by PatchSetPciSerialDeviceRegisterBase()
+};
+STATIC_ASSERT (sizeof (mSerialDevicePmioPort) == sizeof (UINT16), "Unsupported mSerialDevicePmioPort");
+
+STATIC
+CONST UINTN
+mInOutMaxDistance = 64;
+
+VOID
+PatchSetPciSerialDeviceRegisterBase (
+  IN  UINTN  RegisterBase
+  )
+{
+  //
+  // FIXME: This is really ugly, make quirks take a context param.
+  //
+  if (RegisterBase <= MAX_UINT16) {
+    DEBUG ((DEBUG_INFO, "OCAK: Registering PCI serial device PMIO port %u\n", RegisterBase));
+    CopyMem (mSerialDevicePmioPort, &RegisterBase, sizeof (UINT16));
+  }
+
+  //
+  // TODO: Add proper MMIO patch.
+  //
+}
+
+STATIC
+EFI_STATUS
+PatchCustomPciSerialPmio (
+  IN OUT PATCHER_CONTEXT    *Patcher
+  )
+{
+  UINTN       Count;
+  UINT8       *Walker;
+  UINT8       *WalkerPmio;
+  BOOLEAN     FoundPmio;
+  UINT8       *WalkerEnd;
+  UINT8       *WalkerTmp;
+
+  ASSERT (Patcher != NULL);
+
+  Count     = 0;
+  Walker    = (UINT8 *) MachoGetMachHeader (&Patcher->MachContext);
+  WalkerEnd = Walker + MachoGetFileSize (&Patcher->MachContext) - mInOutMaxDistance;
+
+  while (Walker < WalkerEnd) {
+    FoundPmio = FALSE;
+
+    if (Walker[0] == mSerialDevicePmioFind[0]
+      && Walker[1] == mSerialDevicePmioFind[1]
+      && (Walker[2] & 0xF8U) == mSerialDevicePmioFind[2]
+      && Walker[3] == mSerialDevicePmioFind[3]) {
+      WalkerPmio = &Walker[2];
+
+      WalkerTmp = Walker + mInOutMaxDistance;
+      while (Walker < WalkerTmp) {
+        //
+        // Locate instruction in (0xEC) or out (0xEE).
+        //
+        if (*Walker == 0xEC || *Walker == 0xEE) {
+          FoundPmio = TRUE;
+          break;
+        }
+
+        ++Walker;
+      }
+    }
+
+    //
+    // Patch PMIO.
+    //
+    if (FoundPmio) {
+      *WalkerPmio++ = (mSerialDevicePmioPort[0] & 0xF8U);
```

Comment:
@joevt I am not sure whether we should `& 0xF8U` for replacement?

Reply:
For PMIO:
For replacement, we want to preserve the 3 least significant bits.
I think we can assume the 3 least significant bits are always zero (this is a property of PMIO base registers?).
Therefore:
```
*WalkerPmio++ = (RegisterBase & 0xff) | (*WalkerPmio & 7);
*WalkerPmio = RegisterBase >> 8;
```

If you don't think the least significant bits are always zero, then:
```
UINT16 port = RegisterBase + (*WalkerPmio & 7);
*WalkerPmio++ = port & 0xff;
*WalkerPmio = port >> 8;
```

For MMIO:
I think we should not do kernel patches for MMIO. I think we should modify the boot-args value - or intercept `PE_parse_boot_argn` and return our calculated value for `mmio_uart` or `pcie_mmio_uart` if the user has indicated in config.plist that we should do so (with CustomPciSerialDevice) and has not provided those values in boot-args themselves.

Reply:
I have no idea to be honest, nor can I test. Could you please experiment and figure out which method is correct?

As for MMIO, I am ok without patching, until someone explicitly requests that.

We would also like to have this PR merged by next Monday’s release. Do you think we can do it? ^^

Reply:
> EDIT - Yes, the 3x LSB must be zero indeed; see https://github.com/acidanthera/audk/blob/d2f79716cb67c0487433b159f882015773dd18b7/MdePkg/Library/BaseIoLibIntrinsic/IoLibMsc.c#L248 (thanks to @mhaeuser)

That's for 32-bit IO ports. And it's only checking the two least significant bits.
PMIO serial registers uses 8-bit IO ports which can have any value for the three least significant bits.

The question was about the three least significant bits for the IO BAR (the base for all the serial registers) https://en.wikipedia.org/wiki/PCI_configuration_space says that IO BAR uses all the bits except the two least significant bits. Therefore, I would use the `+` method instead of the `|` option.

But wait. https://wiki.osdev.org/PCI#Base_Address_Registers says the size of the BAR is equal to two to the power of the number of bits that don't change (except the info bits) when you write 1's to all the bits of the BAR. Since the serial port IO BAR has size 8 (for all the registers), the IO BAR must be a multiple of 8. Therefore the `|` method should also work.

So either method should work. I will test the current patch today.

Reply:
OK, since I do not have any device test, nor do others do most likely - I guess it will be only you who can tell whether the patch works.

By the way, I just figured out that my previous code could be simplified, which is done at https://github.com/acidanthera/OpenCorePkg/pull/331/commits/ff5d7863fe3af0a1bb5fb0767b494d85f83d34c3.

Reply:
I tested https://github.com/acidanthera/OpenCorePkg/commit/ff5d7863fe3af0a1bb5fb0767b494d85f83d34c3
I disabled my xnu patch to make sure only the CustomPciSerialDevice quirk is used.

If I set Serial->Init to false, then OpenCore doesn't override the Pcd's, and CustomPciSerialDevice does a replacement of 0x3f8 with 0x3f8.

If I set Serial->Init to true, then I see serial output from OpenCore and xnu which means the CustomPciSerialDevice quirk works. But I want to be able to set Serial->Init to false to only have serial output in xnu if I'm only debugging xnu.

- Pcd's should always be overridden, regardless of Serial->Init. I think Serial->Init just needs to control calling SerialPortInitialize.
- CustomPciSerialDevice should not do PMIO patch if the base address is the same as the default 0x3f8.
- This message "Registering PCI serial device PMIO port 1016" should use hex (e.g. "0x3f8") for the base address number.

Reply:
In my opinion,  the current behaviour is correct. Leave Init to be true, remove Target serial bit, and you should only see XNU output.

Reply:
> But I want to be able to set Serial->Init to false to only have serial output in xnu if I'm only debugging xnu.

Maybe I'm wrong about setting Serial->Init to false to do serial only in xnu.

The proper way to disable serial output in OpenCore is to unset `OcLog->Options & OC_LOG_SERIAL`. I don't like these types of flags. The individual flags are hidden (only listed in the documentation, not the .plist) and are difficult to check (need to convert to binary to see if the flag is already set) and to change (need to convert back to binary to change the flag). plist files are xml. xml are already verbose and bulky so you're not saving anything by having multiple flags encoded into a single inscrutable number. I would convert these flags into a `dict` but that's a project for another day and this was probably already considered when creating those multi-flag integers.

OcLog calls SerialPortWrite even if SerialPortInitialize wasn't called. This probably shouldn't happen ... unless the point of setting Serial->Init to false is to use the initialization that is performed by some other code outside OpenCore (such as PciSioSerialDxe) or in some other library used by OpenCore (such as OcDebugLibSerial).

OcDebugLibSerial calls SerialPortInitialize even if Serial->Init is false. This doesn't matter since the default DebugLib is OcDebugLibNull and this doesn't change unless you recompile OpenCore.

Reply:
> In my opinion, the current behaviour is correct. Leave Init to be true, remove Target serial bit, and you should only see XNU output.

I'm ok with that. `OcLog->Options & OC_LOG_SERIAL` comes directly from Debug → Target.

Reply:
> CustomPciSerialDevice should not do PMIO patch if the base address is the same as the default 0x3f8.

> This message "Registering PCI serial device PMIO port 1016" should use hex (e.g. "0x3f8") for the base address number.

I fixed these two at https://github.com/acidanthera/OpenCorePkg/pull/331/commits/8b62364e43b758b29492ebd590468db808e5e449. Now that the patch works, let me go back to documents.